src/libsyntax/ast.rs

```diff
@@ -989,32 +1043,48 @@ impl Expr {
         }
     }
 
+    /// Attempts to reparse as `Ty` (for diagnostic purposes).
     pub(super) fn to_ty(&self) -> Option<P<Ty>> {
+        // Attempt to reparse expression.
+        // Since we're working on the AST level, it's possible to not
+        // know the result type of expression yet.
```

Comment:
This sentence looks like some kind of misunderstanding? We are not trying to get the expression's type here, just reinterpret some piece of code (previously parsed as an expression) as a type.